src/liballoc/string.rs

```diff
@@ -618,7 +618,15 @@ impl String {
     /// ```
     #[stable(feature = "rust1", since = "1.0.0")]
     pub fn from_utf16(v: &[u16]) -> Result<String, FromUtf16Error> {
-        decode_utf16(v.iter().cloned()).collect::<Result<_, _>>().map_err(|_| FromUtf16Error(()))
+        let mut ret = String::with_capacity(v.len());
+        for c in decode_utf16(v.iter().cloned()) {
+            if let Ok(c) = c {
+                ret.push(c);
```

Comment:
Would be nice to have a comment explaining why the code works this way instead of the more "obvious" `collect` call.

Basically, what you wrote in the PR should be in the code.

Reply:
Good idea; since it is already rolled up, I can add this comment afterwards, along with some other assorted code adjustments.